### PR TITLE
[expo-camera] ability to modify video bitrate on android

### PR DESCRIPTION
**Pull request by xHeinrich**
# Why
To allow for more fine grained control over the size of video files generated on android. This is especially useful for apps being used for video services to reduce the amount of data transfer required when uploading video.
https://github.com/react-native-community/react-native-camera/issues/687

# How

Most of the code is taken from the below PR on react-native-camera
https://github.com/react-native-community/react-native-camera/pull/2055

# Test Plan

I dont currently have a mac so am unable to test the changes. I will try get my hands on one to compile the changes and ensure its functioning. The code attached should be sufficient to test the changes.

```
this.camera.recordAsync({
    quality: Camera.Constants.VideoQuality['720p'],
    videoBitrate: 1000
})
```

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -22,6 +22,7 @@ This is the log of notable changes to the Expo client that are developer-facing.
 
 - Added `fallbackLabel` option to `LocalAuthentication.authenticateAsync` on iOS which allows to customize a title of the fallback button when the system doesn't recognize the user and asks to authenticate via device passcode. ([#4612](https://github.com/expo/expo/pull/4612) by [@changLiuUNSW](https://github.com/changLiuUNSW))
 - added `native` mode for Android SplashScreen on standalone apps by [@bbarthec](https://github.com/bbarthec) ([#4567](https://github.com/expo/expo/pull/4567))
+- added `videoBitrate` option for camera component on android
```
Comment (bbarthec) [inline CHANGELOG.md]:
```suggestion
- Added `videoBitrate` option for camera on Android by [@xHeinrich] ([#4878])
```

File: docs/pages/versions/v33.0.0/sdk/camera.md [hunk 2]
```diff
@@ -232,6 +232,7 @@ Starts recording a video that will be saved to cache directory. Videos are rotat
   - **maxDuration (_number_)** -- Maximum video duration in seconds.
   - **maxFileSize (_number_)** -- Maximum video file size in bytes.
   - **mute (_boolean_)** -- If present, video will be recorded with no sound.
+  - **videoBitrate (_number_)** -- Android Only. (int greater than 0) This option specifies a desired video bitrate.  For example, 5\*1000\*1000 would be 5Mbps.
```
Comment (bbarthec) [inline docs/pages/versions/v33.0.0/sdk/camera.md]: Please move these changes to https://github.com/expo/expo/blob/master/docs/pages/versions/unversioned/sdk/camera.md

File: packages/expo-camera/android/src/main/java/expo/modules/camera/ExpoCameraView.java [hunk 3]
```diff
@@ -231,6 +232,10 @@ public void record(Map<String, Object> options, final Promise promise, File cach
         profile = CameraViewHelper.getCamcorderProfile(getCameraId(), ((Double) options.get(QUALITY_KEY)).intValue());
       }
 
+      if (options.get(VIDEO_BITRATE_KEY) != null) {
+        profile.videoBitRate = (int) options.get(VIDEO_BITRATE_KEY);
```
Comment (bbarthec) [inline packages/expo-camera/android/src/main/java/expo/modules/camera/ExpoCameraView.java]:
```suggestion
        profile.videoBitRate = ((Double) options.get(VIDEO_BITRATE_KEY)).intValue();
```

File: packages/expo-camera/src/Camera.tsx [hunk 4]
```diff
@@ -50,6 +50,10 @@ function ensureRecordingOptions(options?: RecordingOptions): RecordingOptions {
     recordingOptions.quality = Camera.Constants.VideoQuality[recordingOptions.quality];
   }
 
+  if (recordingOptions.videoBitrate && typeof recordingOptions.videoBitrate !== 'number') {
+    console.warn('Video Bitrate should be a positive integer');
```
Comment (bbarthec) [inline packages/expo-camera/src/Camera.tsx]: Remove that checking. `Typescript` is ensuring proper typings, we do not want to duplicate its work in runtime.

**Review by bbarthec — CHANGES_REQUESTED** [review]
You need to build `expo-camera` and provide newer `build` directory with your changes (our idea is to have transpilled code ready to be used after user install each unimodule):
```
yarn build
```

Good job 💯 overall, but we need to test your changes when ready, so please adjust PR and I will take of it further.

**Comment by xHeinrich** [comment]
Thanks! I've updated all of the requested items and run yarn build however for some reason the sdk test is still failing, is there something specific I should do to resolve that?

**Comment by bbarthec** [comment]
@xHeinrich, I need to apologise to you - I've lost track of this PR and it fell behind master so terribly 🙇 I'd like to merge your PR, but I need to ask you to refresh it a little by resolving conflicts 🙏

**Comment by wolverineks** [comment]
@xHeinrich any plans to fix the merge conflicts?
i could give it a try if you're too busy. im working on a project right now that might benefit from this

**Comment by bbarthec** [comment]
@wolverineks please do, as @xHeinrich seems unreachable for a while now 😢

**Comment by xHeinrich** [comment]
@wolverineks @bbarthec please feel free to update this, last time i tried running yarn build the tests were still failing and couldnt figure out why. You might have better luck.

**Comment by wolverineks** [comment]
ive made the changes, and ive written a simple test in test-suite/tests/Camera.js
but which commands should i run to get the changes?

ive tried
`yarn build` in expo-camera
`expo start` in test-suite

am i heading in the right direction?

**Comment by wolverineks** [comment]
ok, i think i got it now

**Comment by wolverineks** [comment]
dang, i finally got bare-expo to run, but Camera and GL test options are greyed out
any ideas?

**Comment by wolverineks** [comment]
it looks like this commit wrapped the import so that the Camera.js tests never run if there is an error thrown.
the error thrown is:
`TypeError: undefined is not an object (evaluating 'CameraManager.Type)`
im running on my device (Pixel 3), so do i have to start the app differently to get it to import the camera without throwing?

https://github.com/expo/expo/commit/dc814a6b86b23bc80c8ca78b6579f2e3c6b42abc#diff-225580c4eddce6fbc51d58127b6e3219

**Comment by bbarthec** [comment]
@wolverineks, please don't bother about writing tests for this small change as there's hardly any way to actually test in on JS level.
Please take care of conflicts, by running `yarn build` in `packages/expo-camera` and manual adjustments and I'll take care of the rest 💪

**Comment by wolverineks** [comment]
@bbarthec any value to a test like...
```typescript
t.it('records with selected bitrate', async () => {
        const videoBitRate = 8000000; // 8 Mbps
        const duration = 1; // 1 second
        await mountAndWaitFor(<Camera ref={refSetter} style={style} />);
        const { uri } = await instance.recordAsync({
          maxDuration: duration,
          videoBitRate,
        });
        const fileInfo = await FileSystem.getInfoAsync(uri);
        const actualBitRate = (fileInfo.size * 8) / duration;

        t.expect(actualBitRate).toBeLessThan(videoBitRate * 1.25);
        t.expect(actualBitRate).toBeGreaterThan(videoBitRate * 0.75);
      });
```

@xHeinrich did you ever get around to testing this, on a device or otherwise?

**Comment by xHeinrich** [comment]
@wolverineks i manually tested on device

**Comment by bbarthec** [comment]
@wolverineks, I see a value in the test you are suggesting, but if you're having this amount of problems running them correctly I prefer you not to worry about them and ship the feature 😉
Unfortunately I cannot assure you that I would be of any help in this testing matter right now 😞😭

**Comment by wolverineks** [comment]
if you can confirm that `test-suite/tests/Camera.js` runs on `bare-expo`, ill keep chugging away

**Comment by bbarthec** [comment]
@wolverineks, unfortunately at this moment `camera` test are not working in `bare-expo`, so please don't worry about them 😉
However, as you've already prepared test for this new option, please comment them out and provide them in your PR 😃

**Comment by wolverineks** [comment]
thats too bad. oh well, let us know when the tests are running again. thanks

**Comment by limiaspasdaniel** [comment]
@wolverineks any news on this?

**Comment by wolverineks** [comment]
> @wolverineks any news on this?

i stopped working on this when the test suite wouldn't run
i tried again today, but i couldn't get it to run on android or ios

**Comment by makzee** [comment]
I'm trying to get it working. I have hand picked the changes from xHeinrich repo into my repo, now trying to test it. I don't use Mac but have an old Mac machine and setting up the environment there so that I could test it.

**Comment by makzee** [comment]
@bbarthec I have fixed the conflicts and created a new PR (https://github.com/expo/expo/pull/11932).

**Comment by xHeinrich** [comment]
Thanks for the work getting this into master @makzee

**Pull request closed without merging**